File: typescript/tests/commerce_finance_board.test.ts
// commerce_finance_board.test.ts
// Verifies the CircleAI.Commerce.Finance port: invoice issue, payment recording,
// billed = Σ line.Amount × (1 + TaxPct/100), remaining/outstanding, overdue.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryInvoiceBoard,
  CommerceFinanceDomainContext,
  invoice,
  invoiceLine,
  financePayment,
} from "../src/commerce/finance/index";

describe("InMemoryInvoiceBoard", () => {
  it("remainingOn = billed (with tax) − payments; unknown invoice is 0", () => {
    const b = new InMemoryInvoiceBoard();
    b.issue(
      invoice(
        "i1",
        "c1",
        new Date("2026-01-01"),
        new Date("2026-02-01"),
        [invoiceLine("Item", 100, 15), invoiceLine("Item2", 50, 0)],
        "ZAR",
        "Issued",
      ),
    );
    // billed = 100*1.15 + 50*1.0 = 115 + 50 = 165
    assert.equal(b.remainingOn("i1"), 165);
    b.recordPayment(financePayment("p1", "i1", 65, new Date("2026-01-15T00:00:00Z")));
    assert.equal(b.remainingOn("i1"), 100);
    assert.equal(b.remainingOn("ghost"), 0);
  });

  it("totalOutstanding sums remaining across all invoices", () => {
    const b = new InMemoryInvoiceBoard();
    b.issue(invoice("i1", "c1", new Date("2026-01-01"), new Date("2026-02-01"), [invoiceLine("A", 100, 0)], "ZAR", "Issued"));
    b.issue(invoice("i2", "c1", new Date("2026-01-01"), new Date("2026-02-01"), [invoiceLine("B", 200, 0)], "ZAR", "Issued"));
    b.recordPayment(financePayment("p1", "i1", 40, new Date()));
    assert.equal(b.totalOutstanding(), 60 + 200);
  });

  it("markOverdue flips DueDate<asOf non-paid invoices to Overdue; Overdue() lists them", () => {
    const b = new InMemoryInvoiceBoard();
    b.issue(invoice("i1", "c1", new Date("2026-01-01"), new Date("2026-02-01"), [invoiceLine("A", 100, 0)], "ZAR", "Issued"));
    b.issue(invoice("i2", "c1", new Date("2026-01-01"), new Date("2026-02-01"), [invoiceLine("B", 100, 0)], "ZAR", "Paid"));
    b.issue(invoice("i3", "c1", new Date("2026-06-01"), new Date("2026-12-01"), [invoiceLine("C", 100, 0)], "ZAR", "Issued"));
    b.markOverdue(new Date("2026-07-01"));
    assert.deepEqual(
      b.overdue().map((i) => i.invoiceId),
      ["i1"],
    );
    // "Paid" (case-insensitive) is never marked overdue; future-due stays Issued.
    assert.equal(b.get("i2")?.status, "Paid");
    assert.equal(b.get("i3")?.status, "Issued");
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(CommerceFinanceDomainContext.systemPromptSnippet.includes("[DOMAIN: Commerce.Finance]"));
    assert.deepEqual(CommerceFinanceDomainContext.complianceFlags, ["NCA_34_2005", "SARB_aware", "POPIA", "IFRS"]);
    assert.deepEqual(CommerceFinanceDomainContext.suggestedTools, ["cash_flow_model", "spreadsheet", "web_search"]);
  });
});

File: go/tests/commerce_finance_board_test.go
// commerce_finance_board_test.go
//
// Verifies the CircleAI.Commerce.Finance port (commerce_finance_board.go):
// invoice issue/get, payment recording, tax-inclusive remaining balance,
// total outstanding, and mark-overdue (case-insensitive "Paid" skip).

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func finTime(y int, m time.Month, d int) time.Time {
	return time.Date(y, m, d, 0, 0, 0, 0, time.UTC)
}

func TestFinance_IssueGetAndRemainingWithTax(t *testing.T) {
	b := circleai.NewInMemoryInvoiceBoard()
	inv := circleai.Invoice{
		InvoiceId:  "i1",
		CustomerId: "c1",
		IssueDate:  finTime(2026, 7, 1),
		DueDate:    finTime(2026, 7, 31),
		Currency:   "ZAR",
		Status:     "Issued",
		Lines: []circleai.InvoiceLine{
			{Description: "Item A", Amount: circleai.DecimalFromInt(100), TaxPct: 15}, // 115
			{Description: "Item B", Amount: circleai.DecimalFromInt(200), TaxPct: 0},  // 200
		},
	}
	b.Issue(inv)
	if got, ok := b.Get("i1"); !ok || got.CustomerId != "c1" {
		t.Fatalf("get invoice = %+v ok=%v", got, ok)
	}
	// Billed = 115 + 200 = 315, no payments yet.
	if rem := b.RemainingOn("i1"); !rem.Equal(circleai.DecimalFromInt(315)) {
		t.Fatalf("remaining = %s, want 315", rem)
	}
	// Unknown invoice remaining = 0.
	if rem := b.RemainingOn("ghost"); !rem.IsZero() {
		t.Fatalf("unknown remaining = %s, want 0", rem)
	}
	// Record a partial payment of 115.
	b.RecordPayment(circleai.FinancePayment{PaymentId: "p1", InvoiceId: "i1", Amount: circleai.DecimalFromInt(115), AtUtc: finTime(2026, 7, 10)})
	if rem := b.RemainingOn("i1"); !rem.Equal(circleai.DecimalFromInt(200)) {
		t.Fatalf("remaining after payment = %s, want 200", rem)
	}
}

func TestFinance_TaxInclusiveRemainingMatchesDotNet(t *testing.T) {
	// Reference billed amounts from C# `Amount * (decimal)(1 + TaxPct/100.0)`:
	//   79.99 @ 12.5% => 89.98875 ; 33.33 @ 15% => 38.3295 ; 0.01 @ 15% => 0.0115.
	b := circleai.NewInMemoryInvoiceBoard()
	b.Issue(circleai.Invoice{InvoiceId: "i1", Status: "Issued", Lines: []circleai.InvoiceLine{
		{Description: "A", Amount: circleai.NewDecimal(79, 990_000), TaxPct: 12.5},
		{Description: "B", Amount: circleai.NewDecimal(33, 330_000), TaxPct: 15},
		{Description: "C", Amount: circleai.NewDecimal(0, 10_000), TaxPct: 15},
	}})
	// Sum: 89.98875 + 38.3295 + 0.0115 = 128.32975.
	want := circleai.NewDecimal(128, 329_750)
	if rem := b.RemainingOn("i1"); !rem.Equal(want) {
		t.Fatalf("tax-inclusive remaining = %s, want 128.32975", rem)
	}
}

func TestFinance_IssueCopiesLines(t *testing.T) {
	b := circleai.NewInMemoryInvoiceBoard()
	lines := []circleai.InvoiceLine{{Description: "X", Amount: circleai.DecimalFromInt(10), TaxPct: 0}}
	b.Issue(circleai.Invoice{InvoiceId: "i1", Lines: lines, Status: "Issued"})
	lines[0].Amount = circleai.DecimalFromInt(9999)
	if rem := b.RemainingOn("i1"); !rem.Equal(circleai.DecimalFromInt(10)) {
		t.Fatalf("invoice lines not defensively copied: remaining %s", rem)
	}
}

func TestFinance_TotalOutstandingAndOverdue(t *testing.T) {
	b := circleai.NewInMemoryInvoiceBoard()
	b.Issue(circleai.Invoice{InvoiceId: "i1", DueDate: finTime(2026, 6, 1), Status: "Issued", Lines: []circleai.InvoiceLine{{Amount: circleai.DecimalFromInt(100), TaxPct: 0}}})
	b.Issue(circleai.Invoice{InvoiceId: "i2", DueDate: finTime(2026, 6, 15), Status: "paid", Lines: []circleai.InvoiceLine{{Amount: circleai.DecimalFromInt(50), TaxPct: 0}}})
	b.Issue(circleai.Invoice{InvoiceId: "i3", DueDate: finTime(2026, 12, 1), Status: "Issued", Lines: []circleai.InvoiceLine{{Amount: circleai.DecimalFromInt(70), TaxPct: 0}}})

	// Total outstanding = 100 + 50 + 70 = 220 (no payments; remaining counts all).
	if tot := b.TotalOutstanding(); !tot.Equal(circleai.DecimalFromInt(220)) {
		t.Fatalf("total outstanding = %s, want 220", tot)
	}

	// Mark overdue as of Jul 1: i1 (Issued, past due) -> Overdue; i2 is "paid"
	// (case-insensitive) so skipped; i3 not yet due.
	b.MarkOverdue(finTime(2026, 7, 1))
	overdue := b.Overdue()
	if len(overdue) != 1 || overdue[0].InvoiceId != "i1" {
		t.Fatalf("overdue = %+v, want [i1]", overdue)
	}
	if got, _ := b.Get("i2"); got.Status != "paid" {
		t.Fatalf("paid invoice should not be marked overdue, got %q", got.Status)
	}
}
